basemodels/groundtruth.py

```
@@ -0,0 +1,64 @@
+from typing import Union
+from schematics.models import Model
+from schematics.types import StringType, ListType, URLType, BaseType, ModelType, IntType, FloatType, DictType, UnionType
+
+
+class ILBGroundtruth(Model):
+    """
+    Groundtruth file format for `image_label_binary` job type:
+
+    {
+      "https://domain.com/file1.jpeg": ["false", "false", "false"],
+      "https://domain.com/file2.jpeg": ["true", "true", "true"]
+    }
+    """
+    items = BaseType()
+
+    def validate_items(self, data, value):
+        for k, v in value.items():
+            URLType().validate(k)
+            ListType(StringType(choices=["true", "false"])).validate(v)
+
+
+class ILASGroundtruthDatapoint(Model):
+    coords = ListType(UnionType([IntType, FloatType]))
+
+
+class ILASGroundtruth(Model):
+    """
+    Groundtruth file format for `image_label_area_select` job type
+
+    {
+      "datapoints": [
+        {
+          "https://domain.com/file1.jpeg": {
+            "coords": [303, 183]
+          }
+        },
+        {
+          "https://domain.com/file1.jpeg": {
+            "coords": [361, 239]
+          }
+        }
+      ]
+    }
+    """
+    datapoints = ListType(DictType(ModelType(ILASGroundtruthDatapoint)), required=True)
+
+    def validate_datapoints(self, data, value):
+        for datapoint in value:
+            for key in datapoint.keys():
+                URLType().validate(key)
+
+
+def get_groundtruth_model(data: Union[dict, list], request_type: str,
+                          **kwargs) -> Union[None, ILBGroundtruth, ILASGroundtruth]:
+    """
+    Create appropriate groundtruth model based on request_type
+    """
+    if request_type == 'image_label_binary':
+        return ILBGroundtruth({"items": data})
```

Not sure if this one started with a `items` either - were you using those urls in the ticket to confirm the content?

---

There is no `items` field in the groundtruth file itself indeed, I had to wrap actual data into dict with dummy `items` field, because schematics doesn't provide way to describe model with dynamic keys on the top level. Alternatively I can describe entries with schematics and just expose simple `validate(gt, request_type)` which will loop through key,values of gt and validate them appropriately.

---

also approach with a function iterating manually through entries one by one could actually help with "fail early" problem for large gt files, because schematics always goes through all fields to collect all errors, which could be big time waste and passing large json with errors arond

---

Hm I see. might be worth starting to think about the requester validation side of this code, you might find that theres a better way to perform that validation.

> also approach with a function iterating manually through entries one by one could actually help with "fail early" problem for large gt files, because schematics always goes through all fields to collect all errors, which could be big time waste and passing large json with errors arond

True good point, we'll still need to download the whole file, but I wonder if we can stream it and do exactly what you're saying on a partially consumed file.